aodntools/timeseries_products/gridded_timeseries.py

```diff
@@ -0,0 +1,262 @@
+    :return: interpolated dataset
+    """
+
+    with xr.open_dataset(file_name) as nc:
+        site_code = nc.site_code
+        ## get global attributes
+        global_attributes = nc.attrs
+
+        ## get lat/lon
+        longitude_mean = nc.LONGITUDE.mean()
+        latitude_mean = nc.LATITUDE.mean()
+
+        ## in case no depth bins provided, create depth bins to the nearest rounded depth
+        ## if provided, adjust to the min-max registered depth
+        if not depth_bins:
+            depth_bins = make_depth_bins(nc=nc, increment=depth_bins_increment)
+        else:
+            depth_bins = [float(depth) for depth in depth_bins]     # in case depth_bins provided through inline arguments
+            depth_bins = adjust_depth_bins(depth_bins, nc.DEPTH.min(), nc.DEPTH.max())
+
+        ## create empty containers
+        time_0 = pd.to_datetime('1950-01-01T00:00:00')
+        time_min = nc.TIME.values.min()
+        depth_bin_len = len(depth_bins)
+        VoI_binned = xr.DataArray(np.full((depth_bin_len, 1), np.nan), coords=[[time_0], depth_bins],
+                                  dims=['TIME', 'DEPTH'])
+        VoI_ndepths = xr.DataArray(np.full(1, 0), coords=[[time_0]], dims=['TIME'])
+
+        ## group nc by individual timestamps
+        VoI_grouped = nc.groupby('TIME')
+
+
+        for timestamp, (name, group) in enumerate(VoI_grouped):
+            time = [name]
+            n_depths = int(len(group[VoI]))
+
+            if n_depths >= 2:
+                VoI_values = list(group[VoI].values)
+                depth = list(group.DEPTH.values)
+                ## sort depths
+                depth, VoI_values = sort_depths(depth, VoI_values)
+
+                ## check for max separation
+                depth_mask = get_depth_mask(depth_bins=depth_bins, depths=depth, max_separation=max_separation)
+                ## do the interpolation
+                VoI_gridded = np.interp(depth_bins, depth, VoI_values, left=np.nan, right=np.nan)
+                ## set masked depth bins to zero
+                VoI_gridded = VoI_gridded * depth_mask
+                VoI_gridded[VoI_gridded == 0] = np.nan
+
+            else:
+                VoI_gridded = np.full((depth_bin_len, 1), np.nan)
+
+            VoI_temp = xr.DataArray(VoI_gridded.reshape(depth_bin_len, 1), coords=[time, depth_bins],
+                                    dims=['TIME', 'DEPTH'])
+            VoI_ndepths_temp = xr.DataArray([n_depths], coords=[time], dims=['TIME'])
+
+            VoI_binned = xr.concat([VoI_binned, VoI_temp], dim='TIME')
+            VoI_ndepths = xr.concat([VoI_ndepths, VoI_ndepths_temp], dim='TIME')
+
+    VoI_interpolated = xr.Dataset({VoI: VoI_binned.astype(float),
+                                   VoI + '_count': VoI_ndepths.astype(int)})
+
+    ## drop the very first record as it is dummy
+    VoI_interpolated = VoI_interpolated.where(VoI_interpolated.TIME >= time_min, drop=True)
```

Comment:
this overwrites could be a function

Reply:
Also, if you are masking based on time_min, it's memory wise to apply that before interpolation.

Reply:
it is just for removing the first value of all variables that have been set to a dummy, as I cannot concatenate on an empty dataset in the first iteration

Reply:
Then, just put an 

```
if timestamp == 0:
    Voi_binned = ...
else:
    Voi_binned = xr.concat ...
```



Also, I just noticed that you are using ``` with xr.open_dataset as nc```. I tend to prefer to use ```with``` for short blocks otherwise you affect readability because of the indentation.
If you use just ```nc = xr.open_dataset``` you got a better readability of the code.